### ISSUE TITLE
Allow per-container privacy settings

### ISSUE THREAD

**Issue by amluto** [mozilla/testpilot-containers#318]
The ability to enable first-party isolation per container would be fantastic.  I can imagine other settings being useful, too, e.g. tracking protection.

**Comment by rikkit** [comment]
This would be a great feature. "Private browsing mode" could then just be a specialised undeletable container.

**Comment by groovecoder** [comment]
You're right that PBM is, in a sense, a container in which tracking protection is enabled, cookies and history are deleted on close, etc.

**Comment by biva** [comment]
In addition, like suggested in https://discourse.mozilla-community.org/t/custom-history-settings-for-each-container/14485/3 , it would be great to delete history / cookies / connections per container

**Comment by kjozwiak** [comment]
> In addition, like suggested in https://discourse.mozilla-community.org/t/custom-history-settings-for-each-container/14485/3 , it would be great to delete history / cookies / connections per container

[Bug#1199470](https://bugzilla.mozilla.org/show_bug.cgi?id=1199470) would address clearing history per container.

**Comment by imyxh** [comment]
> In addition, like suggested in https://discourse.mozilla-community.org/t/custom-history-settings-for-each-container/14485/3 , it would be great to delete history / cookies / connections per container

Maybe at the top of the history sidebar and privacy settings page there could be a dropdown menu, with something like:
<kbd>Edit settings for: ▼</kbd>
&nbsp;&nbsp;&nbsp;&nbsp;<kbd>Default container&nbsp;</kbd>
&nbsp;&nbsp;&nbsp;&nbsp;<kbd>Personal (example)</kbd>
&nbsp;&nbsp;&nbsp;&nbsp;<kbd>Shopping (example)</kbd>
etc. I suppose this is really a pointless suggestion you would've arrived at anyway, but hey.
Of course, "edit settings for" could be replaced with "showing history for" or whatever else is relevant.

**Comment by b0urb0n** [comment]
As @biva mentioned, the ability to remove cookies and history on a per-container basis would be very useful.  Currently, there is no way to individually "clean-up" a contaminated container.

**Comment by biva** [comment]
I don't know which issue should remain, as there are probably dupplicates between this current one, and https://github.com/mozilla/multi-account-containers/issues/47 and https://github.com/mozilla/multi-account-containers/issues/48
I understand that there are 2 needs:
- have different settings per container (this current issue)
- be able to clear navigation data per container (including history https://github.com/mozilla/multi-account-containers/issues/47, cookies and all other stuff accesible with Ctrl+Shift+Delete)
@jonathanKingston : which issues should we keep to have 1 (or 2) common place to discuss?

**Comment by captn3m0** [comment]
In addition to the above cookie settings, I'd also like to see a "disable javascript" setting for a container. I'd want to disable javascript (and cookies) on my Default container. Any application that requires either can go into my Personal/Work containers.

**Comment by ArchangeGabriel** [comment]
@captn3m0 That could be done with uMatrix or NoScript + #309.

**Comment by rodneyrod** [comment]
@ArchangeGabriel Is it possible to have per account settings for individual plugins?

**Comment by ArchangeGabriel** [comment]
@rodneyrod Not yet, that is exactly what #309 is about.

**Comment by geotheory** [comment]
Another option I'd love to see that might prove popular would be a container policy to accept cookies for all sites but delete them **as soon as the site's last tab is closed** (vs when browser is closed).

**Comment by captn3m0** [comment]
@geotheory I've been using Cookie Auto Delete extension, which does pretty much the same thing on a per-container basis.

**Comment by willlma** [comment]
Not seeing any progress here. Is this something maintainers would accept a PR for?

**Comment by Dannniello** [comment]
> @geotheory I've been using Cookie Auto Delete extension, which does pretty much the same thing on a per-container basis.

Unfortunately Cookie AutoDelete cannot remove localStorage per container. Only "old fashioned" cookies could be removed per container...

It is better than nothing, but nowadays many sites are using cookies and localStorage in the same time. In very short: localStorage is something like cookies, but could store much more data.

Firefox is not supporting removing localStorage per container - it means that you could remove all localStorage or nothing (for example removing youtube.com localStorage in "Default" container will also remove YouTube settings in "Google YouTube" container).

https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/wiki/Documentation#enable-localstorage-support

> "The API for localstorage cleaning doesn't appear to support Containers. So if cookies from one container are cleared, then all of the site's localstorage from all containers are deleted regardless of whitelisting rules."
